classes/task/update_object_tags.php

```diff
@@ -0,0 +1,150 @@
+        ];
+
+        foreach ($tasks as $task) {
+            $table->data[$task->get_id()] = [$task->get_id(), $task->get_iteration(), $task->get_status_badge()];
+        }
+
+        return html_writer::table($table);
+    }
+
+    /**
+     * Returns a status badge depending on the health of the task
+     * @return string
+     */
+    public function get_status_badge(): string {
+        $identifier = '';
+        $class = '';
+
+        if ($this->get_fail_delay() > 0) {
+            $identifier = 'failing';
+            $class = 'badge-warning';
+        } else if (!is_null($this->get_timestarted()) && $this->get_timestarted() > 0) {
+            $identifier = 'running';
+            $class = 'badge-info';
+        } else {
+            $identifier = 'waiting';
+            $class = 'badge-info';
+        }
+
+        return html_writer::span(get_string('status:'.$identifier, 'tool_objectfs', $this->get_fail_delay()), 'badge ' . $class);
+    }
+
+    /**
+     * Returns iteration count
+     * @return int
+     */
+    public function get_iteration(): int {
+        return !empty($this->get_custom_data()->iteration) ? $this->get_custom_data()->iteration : 0;
+    }
+
+    /**
+     * Execute task
+     */
+    public function execute() {
+        if (!tag_manager::is_tagging_enabled_and_supported()) {
+            // Site admin should know if this migration is running but the fs doesn't support tagging
+            // (maybe they changed fs mid-run?).
+            throw new moodle_exception('tagging:migration:notsupported', 'tool_objectfs');
+        }
+
+        // Since this adhoc task can requeue itself, ensure there is a fixed limit on the number
+        // of times this can happen, to avoid any accidental runaways.
+        $iterationlimit = get_config('tool_objectfs', 'maxtaggingiterations') ?: 0;
+        $iteration = $this->get_iteration();
+
+        if (empty($iterationlimit) || empty($iteration) || $iterationlimit < 0 || $iteration < 0) {
+            // This should never hit here, if it does something is very wrong.
+            // Throw exception so it causes a retry and alerts.
+            throw new moodle_exception('tagging:migration:invaliditerations', 'tool_objectfs');
+        }
+
+        if ($iteration > $iterationlimit) {
+            // Generally this means the site has too many objects or not enough configured iterations.
+            // Regardless it should throw an exception to get the site admins attention.
+            throw new moodle_exception('tagging:migration:limitreached', 'tool_objectfs', '', $iteration);
+        }
+
+        $fs = get_file_storage()->get_file_system();
+
+        // This is checked above in tag_manager::is_tagging_enabled_and_supported, but as a sanity check
+        // ensure this specific method is defined.
+        if (!method_exists($fs, "push_object_tags")) {
+            throw new coding_exception("Filesystem does not define push_object_tags");
+        }
```

Comment:
We still want this to retry for e.g. if there was a blib in the connectivity and the s3 api failed for some reason (but then came good later). Not sure if its worth it to try micro optimise this, I think its ok if it fails (even if unrecoverable) since the intention is its a migration process that is watched actively not something triggered automatically.

Also checked and the fail delay remains unchanged with `MDL-79128` - the faildelay will still increase it will just stop trying to execute it after the `attemptsavailable` are exhausted.